warehouse/query-core/src/main/java/datawave/query/attributes/UniqueFields.java

```diff
@@ -0,0 +1,343 @@
+package datawave.query.attributes;
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonValue;
+import com.google.common.collect.Multimap;
+import com.google.common.collect.Sets;
+import com.google.common.collect.SortedSetMultimap;
+import com.google.common.collect.TreeMultimap;
+import datawave.query.Constants;
+import datawave.query.jexl.JexlASTHelper;
+import org.apache.commons.lang.StringUtils;
+
+import java.io.Serializable;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * Represents a set of fields that have been specified within a {@code #unique()} function, as well any specified granularity levels for each individual field
+ * that should be used when determining uniqueness for the values of each field within a {@link Document}. An instance of {@link UniqueFields} can easily be
+ * captured as a parameter string using {@link UniqueFields#toString()}, and transformed back into a {@link UniqueFields} instance via
+ * {@link UniqueFields#from(String)}.
+ */
+public class UniqueFields implements Serializable {
+    
+    private Multimap<String,UniqueGranularity> fieldMap;
+    
+    /**
+     * Returns a new {@link UniqueFields} parsed from this string. The provided string is expected to have the format returned by
+     * {@link UniqueFields#toString()}. Any fields not specified with a {@link UniqueGranularity} name will be added with the default ALL granularity. All
+     * whitespace will be stripped before parsing. See below for certain edge cases:
+     * <ul>
+     * <li>Given null, null will be returned.</li>
+     * <li>Given an empty or blank string, an empty {@link UniqueFields} will be returned.</li>
+     * <li>Given {@code field1[],field2[DAY]}, or {@code field1,field2[DAY]}, or {@code field1[ALL],field2[DAY]}, a {@link UniqueFields} will be returned where
+     * field1 is added with {@link UniqueGranularity#ALL}, and field2 is added with {@link UniqueGranularity#TRUNCATE_TEMPORAL_TO_DAY}.</li>
+     * </ul>
+     * 
+     * @param string
+     *            the string to parse
+     * @return the parsed {@link UniqueFields}
+     */
+    @JsonCreator
+    public static UniqueFields from(String string) {
+        if (string == null) {
+            return null;
+        }
+        // Strip whitespaces.
+        string = StringUtils.deleteWhitespace(string);
+        
+        if (string.isEmpty()) {
+            return new UniqueFields();
+        }
+        
+        UniqueFields uniqueFields = new UniqueFields();
+        
+        final int finalIndex = string.length() - 1;
+        int currentIndex = 0;
+        String field = null;
+        
+        while (currentIndex < finalIndex) {
```

Comment:
I feel this code could be quite a bit simpler.  Why not simply split on commas (StringUtils.split(string, ',')) and then parse each field separately.  This code I think will work but is a little difficult to maintain.

Reply:
I've simplified it to some degree, but I can't rely entirely on `StringUtils.split(string,',')` due to some edge cases. Let me know if you think it needs additional work.

Reply:
ah right, you could have a comma inside the brackets.... I see now